# Problem 1
Solve the following problem using KKT conditions
$$Min~~f = 4x_1 - 3x_2 + 2x_1^2 - 3x_1x_2 + 4x_2^2$$
$$g_1(x):~~2x_1 - 1.5x_2 = 5$$

The KKT conditions can be written as:
\begin{align}
\frac{\partial f}{\partial x_1} - \lambda \frac{\partial g_1}{\partial x_1} = 0\\
\frac{\partial f}{\partial x_2} - \lambda \frac{\partial g_1}{\partial x_2} = 0\\
g_1(x) - b_1 = 0
\end{align}
which evaluates to:
\begin{align}
4x_1 - 3x_2 - 2\lambda = -4\\
-3x_1 + 8x_2 + 1.5\lambda = 3\\
2x_1 - 1.5x_2 - 5 = 0
\end{align}
This can be solved using a system of linear equations


In [115]:
A = [4 -3 -2; -3 8 1.5; 2 -1.5 0];
b = [-4; 3; 5];
c = A\b
println(c)

fun(x1, x2) = 4x1 -3x2 + 2*x1^2 - 3x1*x2 + 4 * x2^2
opt = fun(c[1], c[2])
println("Optimum: ", opt)

[2.5, 0.0, 7.0]
Optimum: 22.5


The Values then are:
\begin{align}
x_1 = 2.5\\
x_2 = 0\\
\lambda = 7
\end{align}
which gives the objective
$$ f = 22.5$$


### (b) Change the constraint to be:
$$g_1(x) = 2x_1 - 1.5x_2 = 5.1$$
because the change in the constraint is 0.1 we expect that the change in the objective will be $0.1 * \lambda = 0.7$

In [116]:
A = [4 -3 -2; -3 8 1.5; 2 -1.5 0];
b2 = [-4; 3; 5.1];
c2 = A\b2
print(c2)

fun(x1, x2) = 4x1 -3x2 + 2*x1^2 - 3x1*x2 + 4 * x2^2
opt2 = fun(c2[1], c2[2])
println("\nNew optimum: ", opt)
delta_opt = opt2 - opt
println("Change in optimum: ", delta_opt)

[2.55, 0.0, 7.1]
New optimum: 22.5
Change in optimum: 0.7049999999999983


As we can see the change in the optimum value was 0.705 which is very close to the predicted value of 0.7. This shows that the $\lambda$ value accurately predicts the change in the optimum.

### (c) Are the KKT equations for a problem with quadratic objective and a linear equality constraint always linear? Is this true for a problem with a quadratic objective and a linear inequality constraint?

If the problem has a quadratic objective and a linear equality constrant then the KKT equations will be linear,
if a linear inequality constraint is present than the problem will also be linear.

## Problem 3: Solve the following problem using KKT conditions

$$Min~f(x) = x_1^2 + 2x_2^2 + 3x_3^2$$
$$g_1(x): x_1 + 5x_2 = 12$$
$$g_2(x): -2x_1-x_2 - 4x_3 \leq -18$$

First we must formulate the problem to match the required format. We change $g_2$ to be:
$$g_2(x): 2x_1 - x_2 + 4x_3 - 18 \leq 0$$
which in turn formulates the system of linear equations representing the KKT Conditions:
\begin{align}
\begin{bmatrix}
 2&  0&  0&  -1& 2\\ 
 0&  4&  0&  -5& -1\\ 
 0&  0&  6&  0& 4\\ 
 2&  -1&  4&  0& 0\\ 
 1&  5&  0&  0& 0
\end{bmatrix} * 
\begin{bmatrix}
x_1 \\
x_2 \\
x_3 \\
\lambda_2 \\
\lambda_1 
\end{bmatrix} = 
\begin{bmatrix}
0\\0\\0\\18\\12
\end{bmatrix}
\end{align}

In [124]:
A = [2 0 0 -1 2
     0 4 0 -5 -1
     0 0 6 0 4
     2 -1 4 0 0
     1 5 0 0 0]
b = [0
     0
     0
     18
     12]
x = A\b
println("The values of the x vector are: $x")
g2(x1, x2, x3) = 2*x1 - x2 + 4x3 - 18
g1(x1, x2) = x1 + 5x2 - 12
g1_x = g1(x[1], x[2])
g2_x = g2(x[1], x[2], x[3])
println("g1 evaluates to: $g1_x")
println("g2 evaluates to: $g2_x")

The values of the x vector are: [4.71698, 1.4566, 2.50566, 1.91698, -3.75849]
g1 evaluates to: 3.907985046680551e-14
g2 evaluates to: 3.552713678800501e-15


Here we can see that:
$$x_1 = 4.717\\
x_2 = 1.457\\
x_3 = 2.506\\
\lambda_1 = -3.7585\\
\lambda_2 = 1.9170$$
We also note from the code that both of the constraints are binding.  This satisfies the neccessary conditions for an optimum

To check the sufficient conditions we must check that $\nabla_x^2L(\textbf{x}^*, \lambda^*)$ is positive definite

$$\nabla_x^2L(\textbf{x}^*, \lambda^*) = \begin{bmatrix} 1&0&0\\ 0&4&0\\ 0&0&6 \end{bmatrix}
 - \lambda_1 \begin{bmatrix} 0&0&0\\0&0&0\\0&0&0\end{bmatrix} 
 - \lambda_2 \begin{bmatrix} 0&0&0\\0&0&0\\0&0&0\end{bmatrix} = \begin{bmatrix} 1&0&0\\ 0&4&0\\ 0&0&6 \end{bmatrix}$$

In [126]:
L = [1 0 0
    0 4 0
    0 0 6]
isOpt = isposdef(L)
if isOpt
    println("The values correspond to a constrained optimum")
end

The values correspond to a constrained optimum


because the hessian of the lagrangian function is postive definite then we can conclude that the point is a constrained minimum.

## Problem 4
$$Min~f(x) = x_1^2 + x_2^2\\
g_1(x)=x_1^2+x_2^2 - 9 = 0\\
g_2(x)=x_1+x_2^2-1\leq 0\\
g_3(x)=x_1+x_2-1\leq 0$$
Once again the inequality constraints must be reformatted
$$g_2(x)=-x_1-x_2^2+1\geq 0\\
g_3(x)=-x_1-x_2+1\geq 0$$

### (a) Verify that [-2.3723, -1.8364] is a local optimum
Check which constraints are binding

In [119]:
x = [-2.3723 -1.8364]
g1(x) = x[1]^2 + x[2]^2 - 9
g2(x) = x[1] + x[2]^2 - 1
g3(x) = x[1] + x[2] - 1
g1x = g1(x)
g2x = g2(x)
g3x = g3(x)
println("g1 = $g1x")
println("g2 = $g2x")
println("g3 = $g3x")

g1 = 0.00017225000000031798
g2 = 6.49600000000028e-5
g3 = -5.2087


within acceptable roundoff g1 and g2 are binding constraints.  We must now check if the KKT conditions are satisfied.

In [128]:
using NLsolve
function f!(F, x)
    F[1] = 2*x[1] - x[3] * 2.0 * x[1] + x[4]
    F[2] = 1.0 - x[3] * 2.0 * x[2] - x[4] * -2 * x[2]
    F[3] = x[1]^2.0 + x[2]^2.0 - 9.0
    F[4] = -x[1] - x[2]^2.0 + 1.0
end

x0 = [-2.37; -1.836; -1.0; -1.0]
sol = nlsolve(f!, x0)
λ₁ = sol.zero[3]
λ₂ = sol.zero[4]
println("λ₁ = $λ₁")
println("λ₂ = $λ₂")

λ₁ = 0.7785253137160885
λ₂ = 1.0508005262435787


Because both $\lambda$ values are positive the KKT conditions are satisfied. We now check the sufficient conditions.

In [121]:
hessf = [2 0; 0 0]
hessg1 = [2 0; 0 2]
hessg2 = [0 0; 0 -2]
lagrangian = hessf - λ₁ * hessg1 - λ₂ * hessg2
println(lagrangian)

isposdef(lagrangian)

[0.442949 0.0; 0.0 0.54455]


true

Because the lagrangain function is positive definite the point is a constrained optimum

### Verify that [-2.5, -1.6583] is not a local optimum

In [122]:
x = [-2.5, -1.6583]
g1x = g1(x)
g2x = g2(x)
g3x = g3(x)
println("g1 = $g1x")
println("g2 = $g2x")
println("g3 = $g3x")

g1 = -4.1109999999733304e-5
g2 = -0.7500411099999997
g3 = -5.1583000000000006


The only binding constraint is g1, but all of the constraints are feasible.
we now solve for $\lambda_1$
$$2x_1 - \lambda_1(2x_1) = 0\\
\lambda_1 = -2\\
and\\
1 - \lambda_1(2x_2) = 0\\
\lambda_1 = 3.3$$
because we cannot solve for a value of $\lambda_1$ the point is not a local optimum

### Drop the equality constraint from the problem. Using the countour plot above to see where the optimum lies, solve for the optimum using the KKT conditions.
We see that only $g_2$ is binding the system of equations that we need to solve is then
$$2x_1 + \lambda_2(1) = 0\\
1 + \lambda_2(2x_2) = 0\\
-x_1 - x_2^2 +1 = 0$$

In [123]:
function f!(F, x)
    F[1] = 2*x[1] + x[3]
    F[2] = 1.0 - x[3] * -2 * x[2]
    F[3] = -x[1] - x[2]^2.0 + 1.0
end

x0 = [-2.37; -1.836; -1.0]
sol = nlsolve(f!, x0)
x1 = sol.zero[1]
x2 = sol.zero[2]
λ₁ = sol.zero[3]
println("x₁ = $x1")
println("x₂ = $x2")
println("λ₁ = $λ₁")
println("The potential optimum point is [$x1, $x2]")

lagr = hessf - λ₁ * hessg2
isOpt = isposdef(lagr)
if isOpt
    println("The lagrangian is positive definite therefore the potential optimum is a constrained optimum")
end

x₁ = -0.2258029814778883
x₂ = -1.1071598716887676
λ₁ = 0.4516059629557766
The potential optimum point is [-0.2258029814778883, -1.1071598716887676]
The lagrangian is positive definite therefore the potential optimum is a constrained optimum


The point [-0.2258, -1.1072 is the constrained optimum for the problem without the equality constraint.